mart/attack/adversary.py

```diff
@@ -150,35 +170,35 @@ def configure_gradient_clipping(
                 self.gradient_modifier(group["params"])
 
     @silent()
-    def forward(self, *, model=None, **batch):
-        batch["model"] = model
-
-        # Adversary can live within a sequence of model. To signal the adversary should
-        # attack, one must pass a model to attack when calling the adversary. Since we
-        # do not know where the Adversary lives inside the model, we also need the
-        # remaining sequence to be able to get a loss.
-        if model:
-            self._attack(**batch)
+    def forward(self, *, batch: torch.Tensor | list | dict, model: Callable):
+        input, target = batch
 
-        perturbation = self.perturber(**batch)
-        input_adv = self.composer(perturbation, **batch)
+        if self.model_transform is not None:
+            model = self.model_transform(model)
 
-        # Enforce constraints after the attack optimization ends.
-        if model:
-            self.enforcer(input_adv, **batch)
-
-        return input_adv
-
-    def _attack(self, *, input, **batch):
-        batch["input"] = input
+        # Optimization loop only sees the transformed input in batches.
+        batch_transformed = {
+            "input": input,
+            "target": target,
+            "model": model,
+        }
 
         # Configure and reset perturbation for current inputs
         self.perturber.configure_perturbation(input)
 
         # Attack, aka fit a perturbation, for one epoch by cycling over the same input batch.
         # We use Trainer.limit_train_batches to control the number of attack iterations.
         self.attacker.fit_loop.max_epochs += 1
-        self.attacker.fit(self, train_dataloaders=cycle([batch]))
+        self.attacker.fit(self, train_dataloaders=cycle([batch_transformed]))
```

Comment:
Maybe instead of transforming the batch, you should just add a `configure_model` method? That will set the current model as state in `Adversary`.

Reply:
I would like to keep the `batch_transform`, because it is helpful when dealing with normalized input. (e.g. Anomalib normalizes input).

Reply:
Good idea. `configure_model()` is added in 000ce3c1bfaaf350d65f586b0c7767091a723b3a

Reply:
I just realize that it is often a bad idea to assign a `LightningModule` as a member of `LightningModule`, because the lightning Trainer logic could interfere with each other. I am going to move model back into the batch.